--- src/http/ngx_http_request.c	2011-01-20 08:37:58.000000000 -0200
+++ src/http/ngx_http_request.c	2011-09-30 22:55:09.935845001 -0300
@@ -2332,7 +2332,15 @@
 
 #endif
 
+#if (NGX_HTTP_SSL)
+    if (c->ssl != NULL) {
+        n = SSL_peek(c->ssl->connection, buf, 1);
+    } else {
+        n = recv(c->fd, buf, 1, MSG_PEEK);
+    }
+#else
     n = recv(c->fd, buf, 1, MSG_PEEK);
+#endif
 
     if (n == 0) {
         rev->eof = 1;
@@ -2342,9 +2350,19 @@
         goto closed;
 
     } else if (n == -1) {
-        err = ngx_socket_errno;
+#if (NGX_HTTP_SSL)
+		if (c->ssl != NULL) {
+			err = SSL_get_error(c->ssl->connection, n);
+		} else {
+			err = ngx_socket_errno;
+		}
+
+		if ((err != NGX_EAGAIN) && (err != SSL_ERROR_WANT_READ) && (err != SSL_ERROR_WANT_WRITE)) {
+#else
+		err = ngx_socket_errno;
 
-        if (err != NGX_EAGAIN) {
+		if (err != NGX_EAGAIN) {
+#endif
             rev->eof = 1;
             c->error = 1;
 
